### PR TITLE
[SD-1844] Fix error where moving fs mounts only updated configuration

### PR DESCRIPTION
**Pull request by wemrysi**
Moving a filesystem mount was being done directly via the `MountConfigs` effect instead of using the `Mounting` effect. The result was that the mount configuration would be updated, but there would be no change in the actual mounted filesystems.

The fix was to revert back to using `Mounting#remount` to move the mount.

In order to expose the problem, the mount service specs were refactored to verify the state of the mounted filesystems as well as their configuration.

### CODE REVIEW

File: web/src/test/scala/quasar/api/services/MountServiceSpec.scala [hunk 1]
```diff
@@ -24,10 +24,10 @@ import quasar.effect.KeyValueStore
 import quasar.fp._
 import quasar.fp.free._
 import quasar.fs._, PathArbitrary._
-import quasar.fs.mount._
+import quasar.fs.mount.{MountRequest => MR, _}
```
Comment (jedesah) [inline web/src/test/scala/quasar/api/services/MountServiceSpec.scala]: What is this conflicting with?

File: web/src/main/scala/quasar/api/services/mount.scala [hunk 2]
```diff
@@ -102,26 +101,16 @@ object mount {
...
-        op)))
-      },
-      EitherT.leftU[String](ApiError.apiError(
+    M: Mounting.Ops[S]
+  ): EitherT[Free[S, ?], ApiError, String] =
+    parse(dstStr).map(sandboxAbs).cata(dst =>
+      M.remount[T](src, dst)
+        .as(s"moved ${printPath(src)} to ${printPath(dst)}")
+        .leftMap(_.toApiError),
+      ApiError.apiError(
```
Comment (jedesah) [inline web/src/main/scala/quasar/api/services/mount.scala]: Not sure I understand how this fixes the problem from a quick glance. But hey! A shorter more correct solution, I will take that! Does this remove a feature or is this just a simplification?

File: web/src/test/scala/quasar/api/services/MountServiceSpec.scala [hunk 3]
```diff
@@ -141,10 +157,11 @@ class MountServiceSpec extends Specification with ScalaCheck with Http4s with Pa
...
               _    <- M.mountView(f, cfg._1, cfg._2).run.flatMap(orFailF)
 
-              resp <- service(Request(uri = pathUri(f)))
+              r    <- service(Request(uri = pathUri(f)))
+              (res, _) = r
+              body <- lift(res.as[Json]).into[Eff]
             } yield {
-              resp.as[Json].unsafePerformSync must_== cfgStr
-              resp.status must_== Status.Ok
+              (body must_== cfgStr) and (res.status must_== Ok)
```
Comment (mossprescott) [inline web/src/test/scala/quasar/api/services/MountServiceSpec.scala]: As a general comment, it would be better if this sort of refactoring was done in at least a separate commit, so we could more easily understand the changes that actually fixed the bug.

File: web/src/test/scala/quasar/api/services/MountServiceSpec.scala [hunk 4]
```diff
@@ -330,18 +377,22 @@ class MountServiceSpec extends Specification with ScalaCheck with Http4s with Pa
...
+                dst   =  parent </> f
+                after <- M.lookup(dst).run
               } yield {
-                resp.as[String].unsafePerformSync must_== s"added ${printPath(parent </> f)}"
-                resp.status must_== Status.Ok
-
-                after must beSome(MountConfig.viewConfig(cfg))
+                (body must_== s"added ${printPath(dst)}")         and
+                (res.status must_== Ok)                           and
+// TODO: This fails due to how we're bypassing Mounting for views, enable and fix.
```
Comment (mossprescott) [inline web/src/test/scala/quasar/api/services/MountServiceSpec.scala]: Is there a task to come back to this after 3.0(.1)?

File: web/src/main/scala/quasar/api/services/mount.scala (hunk 2 above)
Comment (mossprescott) [inline web/src/main/scala/quasar/api/services/mount.scala]: I'm equally mystified. What was this code trying to accomplish by special-casing fileSystemConfigs? You've essentially reverted part of https://github.com/quasar-analytics/quasar/commit/90eef735f9c3fea2e89b9881361e7b9f35d2b89a; did you understand the intention of that change? Why was it not actually necessary?

File: web/src/test/scala/quasar/api/services/MountServiceSpec.scala (hunk 1 above)
Comment (wemrysi) [inline web/src/test/scala/quasar/api/services/MountServiceSpec.scala]: Nothing, just a shorter name.

Comment (jedesah) [inline web/src/test/scala/quasar/api/services/MountServiceSpec.scala]: I think we should avoid renaming imports unless we have a pretty good reason to do so. I also think that in this particular case the new name is on the short (almost impossible to guess) side anyway.

File: web/src/main/scala/quasar/api/services/mount.scala (hunk 2 above)
Comment (wemrysi) [inline web/src/main/scala/quasar/api/services/mount.scala]: Yes, I forgot to mention this pretty much just reverted that commit.

As for why the original change was made, I do not know and I don't think it was necessary. I also don't think `MountConfigs` should appear anywhere in the api as it is an implementation detail of a particular implementation of `Mounting`, i.e. using it assumes how `Mounting` will be interpreted.

As for exactly what was going on, the actual interpreter of `Mounting` does a couple things when mounting: 1) validates the new mount doesn't violate structure (like mounting a filesystem inside another one), 2) does the mount (i.e. connect to mongo, compile sql, etc) and then, if (2) succeeded, updates the configuration with the new mount. The previous code essentially skipped (1) and (2) for filesystem mounts.

File: web/src/test/scala/quasar/api/services/MountServiceSpec.scala (hunk 3 above)
Comment (wemrysi) [inline web/src/test/scala/quasar/api/services/MountServiceSpec.scala]: Agreed, I was a bit cavalier here.

File: web/src/test/scala/quasar/api/services/MountServiceSpec.scala (hunk 4 above)
Comment (wemrysi) [inline web/src/test/scala/quasar/api/services/MountServiceSpec.scala]: https://slamdata.atlassian.net/browse/SD-1859

File: web/src/test/scala/quasar/api/services/MountServiceSpec.scala (hunk 1 above)
Comment (wemrysi) [inline web/src/test/scala/quasar/api/services/MountServiceSpec.scala]: In this case I needed access to `MountRequest` constructors as well as prisms and, while I could have just imported all of `MountRequest._` into scope, I thought this way made it easier to tell where those identifiers were coming from.

File: web/src/main/scala/quasar/api/services/mount.scala (hunk 2 above)
Comment (mossprescott) [inline web/src/main/scala/quasar/api/services/mount.scala]: It seems like at least some of that might call for special treatment in the case of a move (no need to drop the connector and re-test the connection, for example), but I suppose `remount` could be worrying about it at that level. And just to clarify your description, "doing the mount" also involves updating the filesystem interpreter with the new hierarchy, right? Is that what was really causing the crazy behavior seen in 1844?

File: web/src/test/scala/quasar/api/services/MountServiceSpec.scala (hunk 4 above)
Comment (mossprescott) [inline web/src/test/scala/quasar/api/services/MountServiceSpec.scala]: OK, that's good. Now can you explain what it is that we're not able to test right now? Is it just something the test can't verify, or something that's actually not working right?

File: web/src/main/scala/quasar/api/services/mount.scala (hunk 2 above)
Comment (wemrysi) [inline web/src/main/scala/quasar/api/services/mount.scala]: Aye, but if we want to talk about "moving" first-class, then `Mounting` should get a `Move` primitive.

Regarding "doing the mount", yes, in our current implementation it does and that was also being skipped and yes that is what was causing trouble.

File: web/src/test/scala/quasar/api/services/MountServiceSpec.scala (hunk 4 above)
Comment (wemrysi) [inline web/src/test/scala/quasar/api/services/MountServiceSpec.scala]: This asserts that the effect associated with "mounting a view" happens (and also that the view is "mounted" using the `Mounting` api) and it's not so much that we're not able to test this but that it fails and making it pass is a non-trivial effort.

The only reason the service behaves is that our assumption that `Mounting` will be interpreted into `MountConfigs` (among other things) happens to be correct at the moment. If we changed the implementation of `qusar.fs.mount.Mounter` to use some other mechanism, this service wouldn't work.

Comment (wemrysi) [inline web/src/test/scala/quasar/api/services/MountServiceSpec.scala]: @mossprescott Turns out the effort wasn't as bad as I thought as I was able to get all these tests passing as part of the fix for SD-1833.

**Comment by wemrysi** [comment]
Any other comments? What needs to happen for this to be ready?

File: web/src/test/scala/quasar/api/services/MountServiceSpec.scala (hunk 4 above)
Comment (mossprescott) [inline web/src/test/scala/quasar/api/services/MountServiceSpec.scala]: That's good news and makes me much more comfortable merging this.

**Comment by mossprescott** [comment]
I'd love for @drostron to have a chance to look at this, but under the circumstances I say let's go ahead.

LGTM.
